File: binary_search_tree/Binarysearch.c
/*  BINARY SEARCH
    function for carrying out binary search on given array
    - values[] => given sorted array
    - len => length of the array
    - target => value to be searched
*/
int binarySearch(int values[], int len, int target)
{
    int max = (len - 1);
    int min = 0;
    
    int guess;  // this will hold the index of middle elements
    int step = 0;  // to find out in how many steps we completed the search
    
    while(max >= min)
    {
        guess = (max + min) / 2;
        // we made the first guess, incrementing step by 1
        step++;
        
        if(values[guess] ==  target)
        {
            printf("Number of steps required for search: %d \n", step);
            return guess;
        }
        else if(values[guess] >  target) 
        {
            // target would be in the left half
            max = (guess - 1);
        }
        else
        {
            // target would be in the right half
            min = (guess + 1);
        }
    }
 
    // We reach here when element is not 
    // present in array
    return -1;
}
 
int main(void)
{
    int values[] = {13, 21, 54, 81, 90};
    int n = sizeof(values) / sizeof(values[0]);
    int target = 81;
    int result = binarySearch(values, n, target);
    if(result == -1)
    {  
        printf("Element is not present in the given array.");
    }
    else
    {
        printf("Element is present at index: %d", result);
    }
    return 0;
//}

File: algorithms/ar-bsrh/python3/binary_search.py
def binary_search(item_list, item):
    first = 0
    last = len(item_list) - 1
    found = False
    while first <= last and not found:
        mid = (first + last) // 2
        if item_list[mid] == item:
            found = True
        else:
            if item < item_list[mid]:
                last = mid - 1
            else:
                first = mid + 1
    return found


# This code will simply binary search for the given search element and if it is found it will display 'True'
# and of doesn't 'False'. For Example:

print(binary_search([1, 2, 3, 5, 8], 6))
print("")
print(binary_search([1, 2, 3, 5, 8], 5))
print("")
print(binary_search([13, 46, 675, 464, 54, 6, 7, 8, 645, 43, 22, 45, 6, 5], 5))
print("")
print(binary_search([13, 46, 675, 464, 54, 6, 7, 8, 645, 43, 22, 45, 6, 5], 100))
